locality-checker/examples/cockroachdb-statefulset-secure.yaml

```diff
@@ -0,0 +1,340 @@
+        env:
+        - name: POD_NAMESPACE
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.namespace
+        volumeMounts:
+        - name: certs
+          mountPath: /cockroach-certs
+      - image: duskeagle/locality-checker:0.1 # TODO(joel): Move to cockroachdb organization
+        imagePullPolicy: IfNotPresent
+        name: locality-checker
+        volumeMounts:
+        - mountPath: /etc/cockroach-locality
+          name: cockroach-locality
+        env:
+        - name: KUBERNETES_NODE
+          valueFrom:
+            fieldRef:
+              fieldPath: spec.nodeName
+      affinity:
+        podAntiAffinity:
+          preferredDuringSchedulingIgnoredDuringExecution:
+          - weight: 100
+            podAffinityTerm:
+              labelSelector:
+                matchExpressions:
+                - key: app
+                  operator: In
+                  values:
+                  - cockroachdb
+              topologyKey: kubernetes.io/hostname
+      containers:
+      - name: cockroachdb
+        image: cockroachdb/cockroach:v19.2.6
+        imagePullPolicy: IfNotPresent
+        # TODO: Change these to appropriate values for the hardware that you're running. You can see
+        # the amount of allocatable resources on each of your Kubernetes nodes by running:
+        #   kubectl describe nodes
+        # resources:
+        #   requests:
+        #     cpu: "16"
+        #     memory: "8Gi"
+        #   limits:
+            # NOTE: Unless you have enabled the non-default Static CPU Management Policy
+            # and are using an integer number of CPUs, we don't recommend setting a CPU limit.
+            # See:
+            #   https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/#static-policy
+            #   https://github.com/kubernetes/kubernetes/issues/51135
+            #   cpu: "16"
+            #   memory: "8Gi"        
+        ports:
+        - containerPort: 26257
+          name: grpc
+        - containerPort: 8080
+          name: http
+        readinessProbe:
+          httpGet:
+            path: "/health?ready=1"
+            port: http
+            scheme: HTTPS
+          initialDelaySeconds: 10
+          periodSeconds: 5
+          failureThreshold: 2
+        volumeMounts:
+        - name: datadir
+          mountPath: /cockroach/cockroach-data
+        - name: certs
+          mountPath: /cockroach/cockroach-certs
+        - mountPath: /etc/cockroach-locality
+          name: cockroach-locality
+          readOnly: true
+        env:
+        - name: COCKROACH_CHANNEL
+          value: kubernetes-secure
+        command:
+          - "/bin/bash"
+          - "-ecx"
+          # The use of qualified `hostname -f` is crucial:
+          # Other nodes aren't able to look up the unqualified hostname.
+          - exec /cockroach/cockroach start --logtostderr
+            $(cat /etc/cockroach-locality/locality || true)
```

Comment:
This feels a bit strange as it allows the locality container to implicitly fail but this would continue to work if someone commented out the locality init container.

No suggestions, just wondering what your thoughts are?

Reply:
I'm having impression similar to @chrisseto here. Here two concerns it is coming from:

1. it is not visible what exactly flag or parameter is set here and those details can be important at the moment of composing the arguments
2. I have a sense that `cat` would fail with "not found" message in the stderr which is not a huge problem, but still.

Additionally, about the proposed approach:

> For Kubernetes environments where we expect the nodes to always have region and zone labels, such as Cockroach Cloud, we can use the ERROR_ON_MISSING_LABELS environment variable to ensure the locality-checker init container fails and CockroachDB doesn't start.

I would suggest that we treat containers as a building blocks and not one monolith which `ERROR_ON_MISSING_LABELS` implies. If there are no needed labels, the user can remove the init container from the pod. Otherwise, the tool would be better to a part of the main db container.

To extend that idea, it would be reasonable to be generic in the tool itself so that it could be reused in other pods. One approach would be to accept mapping from node labels to local files. At the expense of increased complexity when using its results in the main container, only node labels and our imagination would limits its applications.